File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        queue = deque()
        if root==None:
            return []
        queue.append(root)
        res = []
        level = 0
        
        while queue:
            s = len(queue)
            level += 1
            arr = []
            while s:
                node = queue.popleft()
                
                if node.left:
                    queue.append(node.left)
                    
                if node.right:
                    queue.append(node.right)
                    
                arr.append(node.val)
                s -= 1
            if level%2==0:
                res.append(arr[::-1])
            else:
                res.append(arr)
            
        return res
            

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (root==NULL) return {};
        vector<vector<int>> ans;
        queue<TreeNode*> q;
        q.push(root);
        int cnt = 0;
        while (q.empty()==false){
            int s = q.size();
            vector<int> level;
            while (s--){
                TreeNode* temp = q.front();
                q.pop();
                if (temp->left){
                    q.push(temp->left);
                }
               
                if (temp->right){
                    q.push(temp->right);
                }
                level.push_back(temp->val);
            }
            cnt++;
            if (cnt%2==0){
                reverse(level.begin(),level.end());
            }
            ans.push_back(level);
        }
        return ans;
    }
};